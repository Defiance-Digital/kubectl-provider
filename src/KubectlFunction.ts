import * as path from 'path';

import { KubectlV25Layer } from '@aws-cdk/lambda-layer-kubectl-v25';
import { CfnOutput, Duration } from 'aws-cdk-lib';
import { IVpc, SubnetType } from 'aws-cdk-lib/aws-ec2';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Code, Function, FunctionOptions, Runtime } from 'aws-cdk-lib/aws-lambda';
import { AwsCliLayer } from 'aws-cdk-lib/lambda-layer-awscli';
import { KubectlLayer } from 'aws-cdk-lib/lambda-layer-kubectl';

import { Construct } from 'constructs';

/**
 * The properties for a KubectlFunction construct
 */
export interface KubectlFunctionProps {
  /**
   * Whether to use kubectl v1.25 or not for the lambda layer. Set this to true if your EKS cluster is 1.25+
   *
   * @default false
   */
  readonly usev125?: boolean;
  /**
   * The Role ARN that is to be assumed during the EKS authentication process to access and run commands on the cluster.
   */
  readonly roleArn: string;

  /**
   * The name of the cluster to access.
   */
  readonly clusterName: string;

  /**
   * The VPC where the Kubernetes cluster is
   */
  readonly vpc: IVpc;

  /**
   * Additional parameters to pass to the Lambda function.
   */
  readonly additionalParams?: FunctionOptions;
}

export class KubectlFunction extends Construct {
  public handler: Function;

  constructor(scope: Construct, id: string, props: KubectlFunctionProps) {

    super(scope, id);
    this.handler = new Function(this, 'KubectlProvider', {
      code: Code.fromAsset(path.join(__dirname, 'handlers')),
      handler: 'index.handler',
      runtime: Runtime.PYTHON_3_10,
      environment: {
        CLUSTER_NAME: props.clusterName,
        ROLE_ARN: props.roleArn,
      },
      timeout: Duration.seconds(45),
      vpc: props.vpc,
      vpcSubnets: {
        subnetType: SubnetType.PRIVATE_WITH_EGRESS,
      },
      ...props.additionalParams,
    });
    this.handler.addToRolePolicy(new PolicyStatement({
      actions: ['eks:DescribeCluster'],
      resources: ['*'],
      effect: Effect.ALLOW,
    }));

    this.handler.addToRolePolicy(new PolicyStatement({
      actions: ['sts:AssumeRole'],
      resources: [props.roleArn],
      effect: Effect.ALLOW,
    }));

    this.handler.addLayers(new AwsCliLayer(this, 'AwsCliLayer'));
    if (props.usev125) {
      this.handler.addLayers(new KubectlV25Layer(this, 'KubectlLayer'));

    } else {
      this.handler.addLayers(new KubectlLayer(this, 'KubectlLayer'));
    }


    new CfnOutput(this, 'SecurityGroupForLambda', {
      value: this.handler.connections.securityGroups[0].securityGroupId,
    });
  }
}
