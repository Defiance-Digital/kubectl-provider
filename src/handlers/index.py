import json
import logging
import os
import subprocess

logger = logging.getLogger()
logger.setLevel(logging.INFO)

# these are coming from the kubectl layer
os.environ['PATH'] = '/opt/kubectl:/opt/awscli:' + os.environ['PATH']

outdir = os.environ.get('TEST_OUTDIR', '/tmp')
kubeconfig = os.path.join(outdir, 'kubeconfig')

def handler(event, context):
    logger.info(json.dumps(event))

    # resource properties (all required)
    cluster_name  = os.environ.get('CLUSTER_NAME') # props['ClusterName']
    role_arn      = os.environ.get('ROLE_ARN') #props['RoleArn']

    # "log in" to the cluster
    cmd = [ 'aws', 'eks', 'update-kubeconfig',
        '--role-arn', role_arn,
        '--name', cluster_name,
        '--kubeconfig', kubeconfig
    ]
    logger.info('Running command: %s' % cmd)
    subprocess.check_call(cmd)

    if os.path.isfile(kubeconfig):
        os.chmod(kubeconfig, 0o600)

    return kubectl(event["commands"])


def kubectl(commands):
    maxAttempts = 3
    retry = maxAttempts
    while retry > 0:
        try:
            cmd = ['kubectl', '--kubeconfig', kubeconfig] + commands

            logger.info('Running command: %s' % cmd)
            output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
            logger.info(f'Output: {output}')

            decoded_str = output.decode('utf-8')
            logger.info(f'Decoded string: {decoded_str}')

            # Split the string into individual JSON objects
            json_objects = decoded_str.strip().split('\n')
            logger.info(f'JSON objects: {json_objects}')

            # Parse each JSON object and store in a list
            parsed_json_list = [json.loads(obj) for obj in json_objects]
            logger.info(f'Parsed JSON list: {parsed_json_list}')

            # Print the list of parsed JSON objects
            for log in parsed_json_list:
                logger.info(json.dumps(log), indent=4)

        except subprocess.CalledProcessError as exc:
            output = exc.output
            if b'i/o timeout' in output and retry > 0:
                retry = retry - 1
                logger.info("kubectl timed out, retries left: %s" % retry)
            else:
                raise Exception(output)
        else:
            logger.info(output)
            return output
    raise Exception('Operation failed after %s attempts: %s' % maxAttempts, output)