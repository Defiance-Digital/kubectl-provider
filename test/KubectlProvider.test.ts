import { App, DockerImage, Duration, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Vpc } from 'aws-cdk-lib/aws-ec2';
import { Code, CodeConfig } from 'aws-cdk-lib/aws-lambda';
import { KubectlFunction } from '../src';

let fromAssetMock: any;
let fromBuildMock: any;

describe('LambdaFunction', () => {
  beforeAll(() => {
    // mock the Code calls so tests run quicker
    fromAssetMock = jest.spyOn(Code, 'fromAsset').mockReturnValue({
      isInline: false,
      bind: (): CodeConfig => {
        return {
          s3Location: {
            bucketName: 'my-bucket',
            objectKey: 'my-key',
          },
        };
      },
      bindToResource: () => {
        return;
      },
    } as any);
    fromBuildMock = jest.spyOn(DockerImage, 'fromBuild').mockReturnValue({
      isInline: false,
      bind: (): CodeConfig => {
        return {
          s3Location: {
            bucketName: 'my-bucket',
            objectKey: 'my-key',
          },
        };
      },
      bindToResource: () => {
        return;
      },
    } as any);
  });
  afterAll(() => {
    // restore the Code from mock
    fromAssetMock?.mockRestore();
    fromBuildMock?.mockRestore();
  });
  test('Basic Lambda Snapshot', () => {
    const app = new App();
    const stack = new Stack(app, 'test');

    const testVpc = Vpc.fromVpcAttributes(stack, 'TestVpc', {
      privateSubnetIds: ['s-123'],
      vpcId: 'vpc-123123',
      availabilityZones: ['us-east-1a'],
    });
    new KubectlFunction(stack, 'TestFunction', {
      roleArn: 'somerolearn',
      clusterName: 'somecluster',
      vpc: testVpc,
    });
    const assert = Template.fromStack(stack);

    const resource = assert.findResources('AWS::Lambda::Function');
    expect(resource).toMatchSnapshot();
  });

  test('Additional Params Lambda Snapshot', () => {
    const app = new App();
    const stack = new Stack(app, 'test');

    const testVpc = Vpc.fromVpcAttributes(stack, 'TestVpc', {
      privateSubnetIds: ['s-123'],
      vpcId: 'vpc-123123',
      availabilityZones: ['us-east-1a'],
    });
    new KubectlFunction(stack, 'TestFunction', {
      roleArn: 'somerolearn',
      clusterName: 'somecluster',
      vpc: testVpc,
      additionalParams: { timeout: Duration.seconds(30), memorySize: 512 },
    });
    const assert = Template.fromStack(stack);

    const resource = assert.findResources('AWS::Lambda::Function');
    expect(resource).toMatchSnapshot();
  });
});
