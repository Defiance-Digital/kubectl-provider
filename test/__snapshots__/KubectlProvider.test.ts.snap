// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaFunction Additional Params Lambda Snapshot 1`] = `
Object {
  "TestFunctionKubectlProvider08B8F324": Object {
    "DependsOn": Array [
      "TestFunctionKubectlProviderServiceRoleDefaultPolicyF73A53D5",
      "TestFunctionKubectlProviderServiceRole182EFC99",
    ],
    "Properties": Object {
      "Code": Object {
        "S3Bucket": "my-bucket",
        "S3Key": "my-key",
      },
      "Environment": Object {
        "Variables": Object {
          "CLUSTER_NAME": "somecluster",
          "ROLE_ARN": "somerolearn",
        },
      },
      "Handler": "index.handler",
      "Layers": Array [
        Object {
          "Ref": "TestFunctionAwsCliLayer2ECCD930",
        },
        Object {
          "Ref": "TestFunctionKubectlLayer88A4BB23",
        },
      ],
      "MemorySize": 512,
      "Role": Object {
        "Fn::GetAtt": Array [
          "TestFunctionKubectlProviderServiceRole182EFC99",
          "Arn",
        ],
      },
      "Runtime": "python3.7",
      "Timeout": 30,
      "VpcConfig": Object {
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "TestFunctionKubectlProviderSecurityGroup03954269",
              "GroupId",
            ],
          },
        ],
        "SubnetIds": Array [
          "s-123",
        ],
      },
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`LambdaFunction Basic Lambda Snapshot 1`] = `
Object {
  "TestFunctionKubectlProvider08B8F324": Object {
    "DependsOn": Array [
      "TestFunctionKubectlProviderServiceRoleDefaultPolicyF73A53D5",
      "TestFunctionKubectlProviderServiceRole182EFC99",
    ],
    "Properties": Object {
      "Code": Object {
        "S3Bucket": "my-bucket",
        "S3Key": "my-key",
      },
      "Environment": Object {
        "Variables": Object {
          "CLUSTER_NAME": "somecluster",
          "ROLE_ARN": "somerolearn",
        },
      },
      "Handler": "index.handler",
      "Layers": Array [
        Object {
          "Ref": "TestFunctionAwsCliLayer2ECCD930",
        },
        Object {
          "Ref": "TestFunctionKubectlLayer88A4BB23",
        },
      ],
      "Role": Object {
        "Fn::GetAtt": Array [
          "TestFunctionKubectlProviderServiceRole182EFC99",
          "Arn",
        ],
      },
      "Runtime": "python3.7",
      "Timeout": 45,
      "VpcConfig": Object {
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "TestFunctionKubectlProviderSecurityGroup03954269",
              "GroupId",
            ],
          },
        ],
        "SubnetIds": Array [
          "s-123",
        ],
      },
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`LambdaFunction Uses v1.25 support 1`] = `
Object {
  "TestFunctionKubectlProvider08B8F324": Object {
    "DependsOn": Array [
      "TestFunctionKubectlProviderServiceRoleDefaultPolicyF73A53D5",
      "TestFunctionKubectlProviderServiceRole182EFC99",
    ],
    "Properties": Object {
      "Code": Object {
        "S3Bucket": "my-bucket",
        "S3Key": "my-key",
      },
      "Environment": Object {
        "Variables": Object {
          "CLUSTER_NAME": "somecluster",
          "ROLE_ARN": "somerolearn",
        },
      },
      "Handler": "index.handler",
      "Layers": Array [
        Object {
          "Ref": "TestFunctionAwsCliLayer2ECCD930",
        },
        Object {
          "Ref": "TestFunctionKubectlLayer88A4BB23",
        },
      ],
      "Role": Object {
        "Fn::GetAtt": Array [
          "TestFunctionKubectlProviderServiceRole182EFC99",
          "Arn",
        ],
      },
      "Runtime": "python3.7",
      "Timeout": 45,
      "VpcConfig": Object {
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "TestFunctionKubectlProviderSecurityGroup03954269",
              "GroupId",
            ],
          },
        ],
        "SubnetIds": Array [
          "s-123",
        ],
      },
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;
